homework/index.js

```diff
@@ -1,46 +1,180 @@
+
+function fetchJSON(url, cb) {
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', url);
+  xhr.responseType = 'json';
+  xhr.onload = () => {
+    if (xhr.status >= 200 && xhr.status <= 299) {
+      cb(null, xhr.response);
+    } else {
+      cb(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+    }
+
+    // ref console.log(xhr.response);
+  };
+  xhr.onerror = () => cb(new Error('Network request failed'));
+  xhr.send();
+}
+
+// createAndAppend	A utility function for easily creating and appending HTML elements.
+// It is not likely that you will need to modify createAndAppend().
+
+function createAndAppend(name, parent, options = {}) {
+  const elem = document.createElement(name);
+  parent.appendChild(elem);
+
+  Object.keys(options).forEach(key => {
+    const value = options[key];
+
+    /* testing lines start 
+        console.log(options) // {text: "AngularJS"}
+        console.log(key) // text
+        console.log(value) // AngularJS
+        testing lines end */
+
+    if (key === 'text') {
+      elem.innerHTML = value; //changed 'textContent' to innerHTML
+    } else {
+      elem.setAttribute(key, value);
+    }
+  });
+  return elem;
+}
+
+// main contains the start-up code for the application.
+// Add new functions and modify function main() as you see fit.
+
+function main(url) {
+  fetchJSON(url, (err, repositories) => {
+    if (err) {
+      createAndAppend('div', root, {
+        text: err.message,
+        class: 'alert-error',
+      });
+      return;
+    }
+    // ref: console.log(repositories);
+
+    //SORT repo names in alphabetical order (case-insensitive) & generate obj with REDUCE
+
+    /* REF https://stackoverflow.com/questions/43773092/how-to-sort-objects-by-value
+        let obj = {a: {'name':'Boy'}, b:{'name': 'apple'}, c:{'name':'Dog'}, d:{'name':'cat'}};
+        Object.keys(obj).sort((a, b) => {
+            a = obj[a].name.toLowerCase();
+            b = obj[b].name.toLowerCase();
+            if (a == b) return 0;
+            return a < b ? -1 : 1;
+        })
+        //return ["b", "a", "d", "c"]
+        */
+    let sorted = Object.keys(repositories)
+      .sort((a, b) => {
+        a = repositories[a].name.toLowerCase();
+        b = repositories[b].name.toLowerCase();
+        if (a === b) return 0;
+        return a < b ? -1 : 1;
+      })
+      .reduce((acc, curr, i) => {
+        acc[i] = repositories[curr];
+        return acc;
+      }, {});
+
+    // ref: console.log(sorted);
+
+    // generate ARRAY of repo names
+    for (const repo in sorted) {
+      repoNamesArr.push(sorted[repo].name);
+    }
+
+    // ref:   console.log(repoNamesArr);
+
+    // create list of repo names from Arr to select from
+    repoNamesArr.forEach(repoName => {
+      createAndAppend('option', select, {
+        text: repoName,
+        value: repoNamesArr.indexOf(repoName), // where each option element has the array index of the repository as its value
+      });
     });
```

Comment:
Everything is working here, great job! But... do keep in mind that this function is getting really long. Ask yourself if you could break down each separate concept into its own function. Will help with understandability of your code.